test/testshared/integration/run.go

```diff
@@ -0,0 +1,116 @@
+package integration
+
+import (
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/golangci/golangci-lint/pkg/logutils"
+	"github.com/golangci/golangci-lint/test/testshared"
+)
+
+const testdataDir = "testdata"
+
+func RunTestdata(t *testing.T) {
+	t.Helper()
+
+	RunTestSourcesFromDir(t, testdataDir)
+}
+
+func RunTestSourcesFromDir(t *testing.T, dir string) {
+	t.Helper()
+
+	t.Log(filepath.Join(dir, "*.go"))
+
+	sources := findSources(t, dir, "*.go")
+
+	binPath := testshared.InstallGolangciLint(t)
+
+	cwd, err := os.Getwd()
+	require.NoError(t, err)
+	t.Cleanup(func() { _ = os.Chdir(cwd) })
+
+	err = os.Chdir(dir)
+	require.NoError(t, err)
+
+	log := logutils.NewStderrLog(logutils.DebugKeyTest)
+	log.SetLevel(logutils.LogLevelInfo)
+
+	for _, source := range sources {
+		source := source
+		t.Run(filepath.Base(source), func(subTest *testing.T) {
+			subTest.Parallel()
+
+			rel, err := filepath.Rel(dir, source)
+			require.NoError(t, err)
+
+			testOneSource(subTest, log, binPath, rel)
+		})
+	}
+}
+
+func testOneSource(t *testing.T, log *logutils.StderrLog, binPath, sourcePath string) {
+	t.Helper()
+
+	rc := testshared.ParseTestDirectives(t, sourcePath)
+	if rc == nil {
+		t.Skipf("Skipped: %s", sourcePath)
+	}
+
+	args := []string{
+		"--go=1.22", // TODO(ldez) remove this line when we will run go1.23 on the CI. (related to intrange, copyloopvar)
+		"--disable-all",
+		"--out-format=json",
+		"--max-same-issues=100",
+		"--max-issues-per-linter=100",
+	}
+
+	for _, addArg := range []string{"", "-Etypecheck"} {
```

Comment:
What are we testing here? Running once with default linters and once with default + typecheck (which isn't a linter and cant be switched off)? Given the code we seem to expect the same result but why do we do it?

Reply:
typecheck needs to be disabled inside tests, so typecheck is disabled by default inside tests.
Only internal tests can disable it.
The fact of having tests without typecheck is important, so we have two test cases.